bindings/cs/rl.net.native/rl.net.live_model.cc

```diff
@@ -1,33 +1,56 @@
 {
-    auto managed_callback_local = context->callback;
-    if (managed_callback_local)
+    auto managed_backgroud_error_callback_local = context->background_error_callback;
+    if (managed_backgroud_error_callback_local)
     {
-        managed_callback_local(status);
+        managed_backgroud_error_callback_local(status);
     }
 }
 
-API livemodel_context_t* CreateLiveModel(reinforcement_learning::utility::configuration* config)
-{
-    livemodel_context_t* context = new livemodel_context_t;
-    context->callback = nullptr;
-
-    context->livemodel = new reinforcement_learning::live_model(*config, pipe_managed_callback, context);
-
-    return context;
-}
-
-API void DeleteLiveModel(livemodel_context_t* context)
-{
-    // Since the livemodel destructor waits for queues to drain, this can have unhappy consequences,
-    // so detach the callback pipe first. This will cause all background callbacks to no-op in the
-    // unmanaged side, which maintains expected thread semantics (the user of the bindings)
-    context->callback = nullptr;
-
-    delete context->livemodel;
-    delete context;
+API livemodel_context_t* CreateLiveModel(reinforcement_learning::utility::configuration* config)
+{
+    livemodel_context_t* context = new livemodel_context_t;
+    context->background_error_callback = nullptr;
+    context->trace_logger_callback = nullptr;
+    context->trace_logger_factory = nullptr;
+
+    const auto binding_tracer_create = [context](reinforcement_learning::i_trace** retval,
+      const reinforcement_learning::utility::configuration& cfg,
+      reinforcement_learning::i_trace* trace_logger,
+      reinforcement_learning::api_status* status)
+    {
+      *retval = new reinforcement_learning::binding_tracer(context->trace_logger_callback);
+      return reinforcement_learning::error_code::success;
+    };
+
+    config->set(reinforcement_learning::name::TRACE_LOG_IMPLEMENTATION, reinforcement_learning::value::BINDING_TRACE_LOGGER);
+    context->trace_logger_factory = new reinforcement_learning::trace_logger_factory_t();;
+    context->trace_logger_factory->register_type(reinforcement_learning::value::BINDING_TRACE_LOGGER, binding_tracer_create);
+
+    context->livemodel = new reinforcement_learning::live_model(*config, pipe_backgroud_error_callback, context, context->trace_logger_factory);
+
+    return context;
+}
+
+API void DeleteLiveModel(livemodel_context_t* context)
+{
+    // Since the livemodel destructor waits for queues to drain, this can have unhappy consequences,
+    // so detach the callback pipe first. This will cause all background callbacks to no-op in the
+    // unmanaged side, which maintains expected thread semantics (the user of the bindings)
+    context->background_error_callback = nullptr;
+    context->trace_logger_callback = nullptr;
+
+    delete context->trace_logger_factory;
+    delete context->livemodel;
+    delete context;

```

Comment:
Would a destructor be useful?  Or unique_ptr for pointer variables might be another way to go.

Reply:
To destructor, same comments to constructor. 

However, I am not sure if we should change to unique_ptr. Here all there fields are actually managed by the C# side. Having unique_ptr may have issue because of the scope - correct me if I am wrong.